### ISSUE TITLE
Error message printed first time python3 (version 3.7.0) dynamic library is imported

### ISSUE THREAD

**Issue by mario-grgic** [vim/vim#3117]
I have compiled vim with  --enable-pythoninterp --enable-python3interp and when I execute :python3 1 from vim I get:

```
/must>not&exist/foo:1: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
```

message on first run only.

I am on macOS 10.13.5, Xcode 9.3 and Python 3.7.0.

**Comment by k-takata** [comment]
Oh, `imp` is deprecated since Python 3.4.
https://docs.python.org/3.7/library/imp.html#module-imp
It seems that we need to use `importlib` instead.

**Comment by puremourning** [comment]
This can trivially be reproduced as follows:

* build vim against python 3.7
* run `vim -Nu NONE --cmd 'py3 import sys'`

**Comment by Khalilw1** [comment]
Is there any news on a possible fix ?

**Comment by bfwg** [comment]
For anyone who's bugged by this, I was able to get around on my Mac with
```
brew reinstall vim --with-python@2
```

**Comment by yzlnew** [comment]
@bfwg Just FYI, building vim with `py2` may cause issues with plugins like `python-mode` if using some specific settings relying on `py3`.

**Comment by Khalilw1** [comment]
@yzlnew I think YouCompleteMe would also complain. If I understand the issue correctly this is just a warning from vim on usage of imp. Meaning that at this point nothing should break am I right ?

**Comment by omdv** [comment]
@Khalilw1 no, in my case it actually broke the vim-latex-live-preview plugin.

**Comment by yzlnew** [comment]
@Khalilw1 YCM works just fine with this error message.

**Comment by Khalilw1** [comment]
@yzlnew thanks for the response although I still find it the warning to be obtrusive 😿. Let us hope a fix is upcoming
@omdv I guess a fix will be needed for it some time soon. thank you for updating me

**Comment by moonfruit** [comment]
I think maybe we can simply add vim specific warning filters to ignore this warning.
reference: python doc for [warnings](https://docs.python.org/3/library/warnings.html).
And later replace `imp` with `importlib` for python 3.1 or above.

**Comment by yzlnew** [comment]
@moonfruit `:silent!` can be used to ignore warnings but I'can figure out which command is causing issues, only the function name.

**Comment by yous** [comment]
@yzlnew If you're getting warnings because of YCM, try edit `autoload/youcompleteme.vim`:

``` diff
diff --git a/autoload/youcompleteme.vim b/autoload/youcompleteme.vim
index 597eb020..32461fa9 100644
--- a/autoload/youcompleteme.vim
+++ b/autoload/youcompleteme.vim
@@ -180,7 +180,7 @@ endfunction


 function! s:SetUpPython() abort
-  exec s:python_until_eof
+  silent! exec s:python_until_eof
 from __future__ import unicode_literals
 from __future__ import print_function
 from __future__ import division
```

**Comment by yzlnew** [comment]
@yous It works. Thanks a lot.

**Comment by yous** [comment]
Regardless of installed plugins, silently execute `python3` once on the top of your vimrc:

``` vim
if has('python3')
  silent! python3 1
endif
```

**Comment by benmezger** [comment]
@yous this worked. But why has it worked?

**Comment by yous** [comment]
@benmezger When we execute `:python3` on Vim, [`DoPyCommand()`](https://github.com/vim/vim/blob/ae728edc93fe5a2f8c39fa75b1ae9ed71526858c/src/if_python3.c#L944) is invoked which calls [`Python3_Init()`](https://github.com/vim/vim/blob/ae728edc93fe5a2f8c39fa75b1ae9ed71526858c/src/if_python3.c#L860). `Python3_Init()` calls `PyImport_AppendInittab("vim", Py3Init_vim);`, and [`Py3Init_vim`](https://github.com/vim/vim/blob/ae728edc93fe5a2f8c39fa75b1ae9ed71526858c/src/if_python3.c#L1599) calls `populate_module(vim_module)`, and that's where actually Vim [imports `imp` module](https://github.com/vim/vim/blob/ad6480961080f80a455b2394f27b02935a2ded52/src/if_py_both.h#L6808).

We forced to skip warnings and errors from `:python3 1` and DeprecationWarning is printed only once, so the error message is gone.

**Comment by eirnym** [comment]
I analysed the code and we don't need the `imp.find_module` executed beforehand. For `load_module` we can use instead `importlib.import_module()` without any loss of functionality

**Comment by shoichiaizawa** [comment]
> Regardless of installed plugins, silently execute python3 once on the top of your > vimrc:
> 
> ```vim
> if has('python3')
>   silent! python3 1
> endif
> ```

Adding this above at the top of my `vimrc` file managed to eliminate the `DeprecationWarning: the imp module...` message but I have another problem and get kicked out of Vim because of that; please see the following error message:

```
Vim: Caught deadly signal SEGV

Vim: Finished.

Segmentation fault: 11
```

This seems to occur only when YouCompleteMe is enabled. And this started happening since around Python 3.7.0 was released.

Any YCM user has encountered this problem?

**Comment by andreaswachowski** [comment]
@shoichiaizawa I guess you will simply have to recompile YouCompleteMe after changing the Python distribution. Change to YouCompleteMe's installation directory and execute `./install.py`, then try again starting vim. For details, see https://github.com/Valloric/YouCompleteMe#installation

**Comment by shoichiaizawa** [comment]
@andreaswachowski 

Thanks for the message. I tried reinstalling YCM but it didn't help me. Just in case, I also did `brew reinstall` Vim and Python 3 but they didn't resolve the issue either... 😢

**Comment by puremourning** [comment]
If you are having trouble with YCM please see the first section of the Readme.

**Comment by sethdeckard** [comment]
@shoichiaizawa Your problem is completely unrelated to this issue but in short you need to make sure that you run `./install.py` with the exact same python that vim is compiled with, not just the same version but the exact same installation (in the case you're using pyenv, etc.). You mention you're using homebrew which means vim will likely be compiled with homebrew installed python, try running `/usr/local/bin/python3 ./install.py`.

**Comment by shoichiaizawa** [comment]
@andreaswachowski @puremourning @sethdeckard 

I recompiled YCM with the Homebrew version of Python 3 and now works fine like before. The problem was persistent with the pyenv version of Python 3 (even though the pyenv Python 3 was installed with the `--enable-framework` option and Vim was working fine with it up until recently). I initially thought my problem was related to this issue to some extent as the problem started happening since when Python 3.7.0 was released.

Thanks for leaving the comments anyway, even though it was off topic 🙏

**Comment by neybar** [comment]
@shoichiaizawa Can you explain what you did a little more?  I tried recompiling with the homebrew python3, but I'm still seeing this error.

I tried the trick of firing off python3 at the top of my vimrc, but that feels like I'm just hiding the problem, not actually fixing it.

**Comment by Khalilw1** [comment]
@neybar here is my current understanding of it hope this helps, I guess because of the deprecation a warning is raised since vim still uses imp module the only way to fix is to adhere to the new way which is importlib. There is a PR I saw which fixes the problem I guess it is not yet merged. I think compiling with homebrow python3 won't fix the problem because the problem is because of python 3.7.

**Comment by yous** [comment]
Patch 8.1.0201 has introduced which fixes this issue. https://github.com/vim/vim/commit/79a494d5e2f97c10e74f92ea529552623c314422

Homebrew vim has also updated to 8.1.0202, so the previous fix can be updated to:

``` vim
if has('python3') && !has('patch-8.1.201')
  silent! python3 1
endif
```

**Comment by k-takata** [comment]
Fixed by 8.1.0201, so closing.

@yous You can just use `has('patch-8.1.201')`.
